src/posts/ci-and-flow/index.md

```diff
@@ -0,0 +1,286 @@
+| Create PR       |           | 0                  |
+| Approve changes | 16 h 57 m | 16 h 57 m          |
+| Merge changes   | 0 h 39 m  | 17 h 36 m          |
+
+It is important to note that this does _not_ include the time it takes to write the code -
+it only measures when the developer thinks the code is ready for integration until it has been integrated.
+
+As can be seen above,
+asynchronous teamwork means waiting.
+And waiting either means not doing anything
+or context-switching to something else.
+Both result in lower productivity and higher lead times.
+
+# Code quality
+
+A common crutch when arguing for pull requests is that it is used to uphold code quality and stop bad design from entering the main branch.
+
+If asynchronous code review was the only way of achieving this, then surely we should have higher quality software these days than before GitHub popularised the pull request.
+Yet, I don't recall either extremely buggy software,
+or sending all code changes as patches over email
+(in fact, the overall quality of software seems to be in decline, but that's a topic for a future blog post).
+There must be other ways of achieving this that don't lower productivity.
+
+# Upholding quality through synchronous teamwork
+
+Synchronous teamwork means that we work together at the same time.
+Instead of sending diffs and comments back and forth we collaborate in real time.
+
+There are, generally, two kinds of synchronous collaboration in development: code review and peer programming.
+
+## Synchronous code review
+
+Instead of sending someone a diff and then waiting for them to read it and come back with feedback,
+the more efficient way of conducting code review is to set up a quick meeting (in person or via video call) and then walk through the changes together.
+
+This means that any issues can be resolved immediately,
+and after the review has concluded,
+the changes can be integrated into the main branch.
+Another benefit is that since everyone is present at the same time,
+the author of the changes can guide the reviewer(s) through them.
+
+A typical synchronous review takes between 5 and 15 minutes,
+vastly faster than the asynchronous variant, which can span days.
+
+## Peer programming
+
+Peer programming is a collective term for pair programming or group programming.
+It happens when two or more developers sit down and develop the same changes together - on one screen.
+
+The most important part of peer programming is that everyone involved should focus on the task.
+If you're not currently active in producing the code,
+you pay attention to the process and review code and design as they emerge.
+
+Peering has multiple benefits over solo development.
+Chief among them is that the collaborative process means that code quality is higher and the development gets stuck less often.
+Another benefit is that once the code has been committed,
+it can be integrated straight into main since it was reviewed as it was written.
```

Comment:
Inte säkert. Om två juniorer (eller egentligen handlar det inte nödvändigtvis om junioritet, utan mer graden av bekantskap med kodbasen) parprogrammerar så kan det ändå vara relevant att låta någon annan kolla på resultatet därefter.

Reply:
Det förutsätter ju att du även har regler för dina PRs att en junior inte får godkänna en annan juniors PR, vilket känns väldigt ovanligt - jag har nog aldrig sett det i praktiken.

Reply:
Github har konceptet "Code owner", vilket jag gissar har med detta att göra, men jag har aldrig provat det:

![image](https://github.com/user-attachments/assets/f1889090-fa7d-43fa-aebb-e16998650184)

Däremot har jag granskat kod som en junior skrivit och som granskats av en annan junior, men där granskningen varit bristfällig (dvs jag godkände inte PR:en).

Reply:
Absolut. Men fråntogs den junioren rättigheten att ensamt godkänna efter det?

Jag vet att det finns de som kräver att en senior utvecklare eller arkitekt ska godkänna koden innan den får komma in. Men att det är möjligt är ju inte samma sak som att det är en bra idé 🙂

Om man är orolig för kvaliteten på juniorernas kod kanske de ska para med en senior istället för varandra - det lär med stor sannolikhet snabba på deras utveckling också.